### PR TITLE
[Fix #9792] Fix false positive for `Lint/Void` cop

### LINKED ISSUE #9792: Lint/Void false positive result
## Expected behavior

```ruby
class Foo
  def foo
    1..100.times do |_n|
      break unless rand > 0.5
    end
    raise
  end
end
```

be ok

## Actual behavior

```bash
$ be rubocop
foo.rb:5:5: W: Lint/Void: Literal 1..100.times do |_n|
      break unless rand > 0.5
    end used in void context.
    1..100.times do |_n| ...
    ^^^^^^^^^^^^^^^^^^^^
```

If replace to `(1..100).times` then everything works as it should

## RuboCop version

```
1.14.0 (using Parser 3.0.1.1, rubocop-ast 1.5.0, running on ruby 2.7.2 x86_64-darwin19)
```

### PR DESCRIPTION
**Pull request by tejasbubane**
Closes #9792

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: changelog/fix_lint_void_false_positive.md [hunk 1]
```diff
@@ -0,0 +1 @@
+* [#9792](https://github.com/rubocop/rubocop/issues/9792): Fix false positive for `Lint/Void` cop. ([@tejasbubane][])
```
Comment (koic) [inline changelog/fix_lint_void_false_positive.md:1]: Can you explain a little more about the fix?

File: lib/rubocop/cop/lint/void.rb [hunk 2]
```diff
@@ -104,7 +104,7 @@ def check_var(node)
         end
 
         def check_literal(node)
-          return if !node.literal? || node.xstr_type?
+          return if !node.literal? || node.xstr_type? || node.irange_type?
```
Comment (koic) [inline lib/rubocop/cop/lint/void.rb:107]:
```suggestion
          return if !node.literal? || node.xstr_type? || node.range_type?
```

File: spec/rubocop/cop/lint/void_spec.rb [hunk 3]
```diff
@@ -254,4 +254,13 @@ def foo=(rhs)
       nil
     RUBY
   end
+
+  it 'handles method with irange block' do
```
Comment (koic) [inline spec/rubocop/cop/lint/void_spec.rb:258]: It looks like `accepts` instead of `handles`.

File: spec/rubocop/cop/lint/void_spec.rb [hunk 4]
```diff
@@ -254,4 +254,13 @@ def foo=(rhs)
...
+    expect_no_offenses(<<~RUBY)
+      def foo
+        1..100.times.each { puts 1 }
+        raise
```
Comment (koic) [inline spec/rubocop/cop/lint/void_spec.rb:262]: That's just my two cents. `do_something` may be better as an example than `raise`.

```suggestion
        do_something
```

File: spec/rubocop/cop/lint/void_spec.rb [hunk 5]
```diff
@@ -254,4 +254,13 @@ def foo=(rhs)
...
+      end
+    RUBY
+  end
```
Comment (koic) [inline spec/rubocop/cop/lint/void_spec.rb:265]: Can you add a test for `erange`?

**Comment by tejasbubane** [comment]
@koic Changed as per your comments.

File: spec/rubocop/cop/lint/void_spec.rb [hunk 6]
```diff
@@ -254,4 +254,14 @@ def foo=(rhs)
       nil
     RUBY
   end
+
+  it 'accepts method with range block' do
+    expect_no_offenses(<<~RUBY)
+      def foo
+        1..100.times.each { puts 1 }
+        1...100.times.each { puts 1 }
```
Comment (koic) [inline spec/rubocop/cop/lint/void_spec.rb:262]: Can you split it into irange and erange tests?

Comment (tejasbubane) [reply spec/rubocop/cop/lint/void_spec.rb:270]: Done.

File: lib/rubocop/cop/lint/void.rb [hunk 7]
```diff
@@ -104,7 +104,7 @@ def check_var(node)
         end
 
         def check_literal(node)
-          return if !node.literal? || node.xstr_type?
+          return if !node.literal? || node.xstr_type? || node.range_type?
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/void.rb:107]: I wonder why `literal?` doesn't cover them. I think they weren't immutable or something like this, but my memory on the subject is fuzzy.

**Pull request merged**